src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -343,6 +343,8 @@ private function createTypeAlternatives(TypedReference $reference)
             $message = sprintf('the existing "%s" service', $this->types[$type]);
         } elseif ($reference->getRequiringClass() && !$reference->canBeAutoregistered()) {
             return ' It cannot be auto-registered because it is from a different root namespace.';
+        } elseif (is_array($this->types) && (false !== $key = array_search(strtolower($type), array_map('strtolower', $this->types)))) {
```

Comment:
I think the logic should be improved: this check should come first in the method, and should account for aliases.
It should also *not* rely on `$this->types`, but on the ids+aliases found in the container.

Reply:
Why not `types` ? What's the difference? :)

Reply:
autowiring is based on the **id** of services, not on their type. So if there is a case mismatch, it's at the **id** level, not at the type one.

Reply:
Moving the check to the first of this `if` and using the `$this->container->getServiceIds()` instead of `$this->types` fails with the following example:
> **Symfony\Component\DependencyInjection\Tests\Compiler\AutowirePassTest::testDontUseAbstractServices**
> Cannot autowire service "bar": argument "$foo" of method "Symfony\Component\DependencyInjection\Tests\Compiler\Bar::__construct()" references class "Symfony\Component\DependencyInjection\Tests\Compiler\Foo" but this service is abstract. Do you mean "Symfony\Component\DependencyInjection\Tests\Compiler\Foo" ?

Reply:
I guess we should add a "!$container->has()" check in the "if"